api/saml/controller.py

```diff
@@ -0,0 +1,249 @@
+
+        :param redirect_uri: Redirection URL
+        :type redirect_uri: string
+
+        :param problem_detail: ProblemDetail object describing the error
+        :type problem_detail: ProblemDetail
+
+        :return: Redirection response
+        :rtype: Response
+        """
+        return redirect(self._error_uri(redirect_uri, problem_detail))
+
+    def saml_authentication_redirect(self, params, db):
+        """Redirects an unauthenticated patron to the authentication URL of the
+        appropriate SAML IdP.
+        Over on that other site, the patron will authenticate and be
+        redirected back to the circulation manager, ending up in
+        saml_authentication_callback.
+
+        :param params: Query parameters
+        :type params: Dict
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :return: Redirection response
+        :rtype: Response
+        """
+        provider_name = params.get(self.PROVIDER_NAME)
+        idp_entity_id = params.get(self.IDP_ENTITY_ID)
+        redirect_uri = params.get(self.REDIRECT_URI, request.path)
+
+        provider = self._authenticator.saml_provider_lookup(provider_name)
+        if isinstance(provider, ProblemDetail):
+            return self._redirect_with_error(redirect_uri, provider)
+
+        authentication_manager = self._get_authentication_manager(db, provider)
+        relay_state = self._add_params_to_url(
+            redirect_uri,
+            {
+                self.LIBRARY_SHORT_NAME: provider.library(db).short_name,
+                self.PROVIDER_NAME: provider_name,
+                self.IDP_ENTITY_ID: idp_entity_id
+            }
+        )
+        redirect_uri = authentication_manager.start_authentication(idp_entity_id, relay_state)
+
+        return redirect(redirect_uri)
+
+    def saml_authentication_callback(self, request, db):
+        """Creates a Patron object and a bearer token for a patron who has just
+        authenticated with one of our SAML IdPs
+
+        :param request: Flask request
+        :type request: Request
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :return: Redirection response or a ProblemDetail if the response is not correct
+        :rtype: Union[Response, ProblemDetail]
+        """
+        if self.RELAY_STATE not in request.form:
+            return SAML_INVALID_RESPONSE.detailed('{0} is empty'.format(self.RELAY_STATE))
+
+        relay_state = request.form[self.RELAY_STATE]
+        relay_state_parse_result = urlparse.urlparse(relay_state)
+        relay_state_parameters = urlparse.parse_qs(relay_state_parse_result.query)
+        library_short_name = relay_state_parameters[self.LIBRARY_SHORT_NAME][0]
+        provider_name = relay_state_parameters[self.PROVIDER_NAME][0]
+        idp_entity_id = relay_state_parameters[self.IDP_ENTITY_ID][0]
```

Comment:
Again, we'll have problems if these keys aren't present in the query string.

Reply:
Added additional validation logic to check whether a specific key in the relay's state parameters and return a ProblemDetail instance in the case if it's not